### ISSUE TITLE
commonLabels creates selector for v1beta1 resources

### ISSUE THREAD

**Issue by jgustie** [kubernetes-sigs/kustomize#1567]
I have an application that uses the older `apps/v1beta1` version of `StatefulSet`; for these resources the `spec/selector` is not required and `spec/selector/matchLabels` will default to `spec/template/metadata/labels`. [This behavior](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#statefulsetspec-v1beta1-apps) is also [documented](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#daemonsetspec-v1beta1-extensions) for the `extensions/v1beta1` version of `DaemonSet`.

If a `commonLabel` transformation is applied to such a resource that has a non-empty `spec/template/metadata/labels` it will add the labels but will also create the missing `spec/selector` using only the `commonLabel` labels. This results in an incorrect selector (all of the existing template labels are suddenly dropped); however by simply not creating the selector, the default behavior is restored and everything works (the transformed set of labels becomes the default value for the selector).

My first attempt to work around the issue was to add the following configuration:

```
commonLabels:
- path: spec/selector/matchLabels
  create: false
  group: apps
  version: v1beta1
  kind: StatefulSet
```

This does not create a conflict as seen in #817 because the added version differentiates it from the default configuration.

I thought that because it was "more specific" it would take precedence. When this did not work I also attempted to save the default configuration and use it as a `builtin` transformer with my added field spec appearing first. This didn't work either, I'm pretty sure it is because the GVK sort order considers only the kind before falling back to a sort that places the versioned instance after the un-versioned instance (and since the first matching field spec is used, my field spec was always ignored).

TL;DR: Should the default config for `commonLabel` include a `create: false` field spec specifically for a `apps/v1beta1:StatefulSet`/`extensions/v1beta1:DaemonSet` and should the GVK sort order be adjusted so that GVKs with a version will be come before those without a version so they are considered first?

**Comment by jbrette** [comment]
Have a look here for the [kustomize 3.2 out of the box solution](https://github.com/keleustes/kustomize/tree/allinone/examples/issues/issue_1567_b). Also this currently the only solution (and some member of the community thinks is should stay this way), you will soon see as soon as you start to create staging and production overlays, that is unmaintainable because you have to copy/paste the transformer.yaml in base/staging and production and that you are better off using jsonpatching.

If you use this [PR Skip kustomize transformers for paths](https://github.com/kubernetes-sigs/kustomize/pull/1491) you will see that you will achieve the same result. But when you start to have base, staging, production....you will only have to put the commonlabels.yaml in on place. [Your issue has been reproducted here](https://github.com/keleustes/kustomize/tree/allinone/examples/issues/issue_1567). The PR still needs some work to become user friendly.

For info, I'm not sure you spend much time trying to use apps/v1betax. The community is scrambling to update manifests and charts right now because Kubernetes 1.16 forces you to use apps/v1.

I also updated the test for use case #817

**Comment by jgustie** [comment]
For most of my work I am on the GKE release train so 1.16 is still a ways off ;)

Also, just to be clear: this is not an issue for resources that do not give special meaning to an empty selector. The problem I am concerned with here only arrises when an empty selector is created during the transformation using only the new labels and it does not account for the special meaning of the previously empty selector (i.e. the existing implicit labels).

Rather than try to mimic all of the cases in the API where labels are copied from the templates, I think it's easier to just not create the selector: this is how `Jobs` already work today. Going through the API it looks as if `ReplicationController` also has special meaning for an empty selector and should have `create: false` in the default configuration.

The other problem is more generic with merging and sorting of field specs: especially with multi-version CRDs, I would expect field specs to work with different versions of resources with the same kind (i.e. irrespective of anything to do with labels and selectors, the configuration workaround above should have been sufficient since it was more specific then the defaults).

**Comment by jgustie** [comment]
I opened PR #1569 that I believe address this specific issue.

**Comment by jgustie** [comment]
I just closed #1569, I misdiagnosed what I was seeing. The GVK sort order is correct, I incorrectly thought that only the first matching field spec was applied when in fact all of the matching field specs are applied (so subsequent matches would still erroneously create the selector even if the initial—more specific—match did not).

I still think the there is a larger issue though: `CreateIfNotFound` should be allowed to have different values for different GV combinations. In this case `create=true` for `spec/selector/matchLabels` on a `StatefulSet` with `apps/v1` is valid (the `selector` is required), but for `apps/v1beta1` it is not (an empty selector means pull labels from the template, creating it without respecting those semantics breaks the definition of the object). Even with `v1beta1` not being supported on the latest versions of Kube, the underlying issue is likely to persist, especially in the CRD space (where Kustomize configurations are provided by CRD authors).

**Comment by jbrette** [comment]
@jgustie Can you have a look at this [PR](https://github.com/kubernetes-sigs/kustomize/pull/1491). The issue you are fighting is here since ages. There a lot of similar issues to yours. We attempted to merge your work, @speth-nr work and ours as you can see in the PR commits.

We recreated an environment to reproduce your issue without having to skip anything, just by being able to deal with version/kind properly (I think) [here](https://github.com/keleustes/kustomize/tree/allinone/examples/issues/issue_1567_c)

The key as you highlighted in your PR is to deal with fieldspecs array in more consistent manner. This PR is wrapping the []FieldSpecs into a struct that can be smart in selecting the fieldspecs and not let the transformers blindly iterate over the []FieldSpec and stop as soon as one matches
It would be really cool if you could check the PR solves your issue and if we could continue working on it together. 

The key points are here: [labeltransformer](https://github.com/keleustes/kustomize/blob/fsskip/pkg/transformers/labelsandannotations.go#L81) and here: [FieldSpecs](https://github.com/keleustes/kustomize/blob/fsskip/pkg/transformers/config/fieldspec.go#L210) 

/cc @ian-howell

**Comment by jgustie** [comment]
I have been giving this some more thought and it occurs to me that the "fix" would actually be to just to set `create: false` for all the affected selectors. My reasoning is that with old versions of the resources creating the selector fundamentally changes the structure of the application (since the created selector only matches labels being added) and new versions _already require the selector to be present_. This is more a function of luck with the current state of the Kube API design, but it makes the change much more presentable as a "bug fix".
